book/layout.md

```diff
@@ -5,529 +5,632 @@ prev: html
-borders have width). That\'s going to mean extending the display list
-to draw both rectangles and text. Let\'s create some data structures
-for that.
+The `render` method now needs to run graphics commands on our actual
+canvas. Let's do this with a `draw` method for each command. On
+`DrawText` it calls to `create_text`:
 
 ``` {.python}
 class DrawText:
-    def __init__(self, x, y, text, font):
-        self.x = x
-        self.y = y
-        self.text = text
-        self.font = font
-
-    def draw(self, scrolly, canvas):
+    def draw(self, scroll, canvas):
         canvas.create_text(
-            self.x, self.y - scrolly,
+            self.x1, self.y1 - scroll,
             text=self.text,
             font=self.font,
             anchor='nw',
         )
+```
 
-class DrawRect:
-    def __init__(self, x1, y1, x2, y2):
-        self.x1 = x1
-        self.y1 = y1
-        self.x2 = x2
-        self.y2 = y2
+Note that the `draw` command accepts the scroll amount as a parameter;
+in effect, we're asking each graphics command to perform coordinate
+conversion.
+
+`DrawRect` works the same way, except it calls `create_rectangle`. By
+default `create_rectangle` draws a one-pixel black border, so make
+sure to pass `width = 0`:
 
-    def draw(self, scrolly, canvas):
+``` {.python}
+
+class DrawRect:
+    def draw(self, scroll, canvas):
         canvas.create_rectangle(
-            self.x1, self.y1 - scrolly,
-            self.x2, self.y2 - scrolly,
+            self.x1, self.y1 - scroll,
+            self.x2, self.y2 - scroll,
             width=0,
-            fill="black",
+            fill=self.color,
         )
 ```
 
-Here I\'ve passed `width=0` to `create_rectangle`, because otherwise
-`create_rectangle` would draw a one-pixel black border.
-
-Now `InlineLayout.layout` will create `DrawText` objects, while
-`BlockLayout.display_list` will create `DrawRect` objects for the
-borders.
+We do still want to clip out graphics commands that only occur
+offscreen. `DrawRect` already contains a `y2` field we can use, so
+let's add the same `DrawText`:
 
 ``` {.python}
-def display_list(self):
-    dl = []
+def __init__(self, x1, y1, text, font):
     # ...
-    _ol, _ot = self.x, self.y
-    _or, _ob = _ol + self.w, _ot + self.h
-    _il, _it = _ol + self.bl, _ot + self.bt
-    _ir, _ib = _or - self.br, _ob - self.bb
-    if self.bl: dl.append(DrawRect(_ol, _ot, _il, _ob))
-    if self.br: dl.append(DrawRect(_ir, _ot, _or, _ob))
-    if self.bt: dl.append(DrawRect(_ol, _ot, _or, _it))
-    if self.bb: dl.append(DrawRect(_ol, _ib, _or, _ob))
-    return dl
+    self.y2 = y1 + font.metrics("linespace")
 ```
 
-Here I define the variables `_ol`, `_il`, `_ot`, `_it`, `_or`, `_ir`,
-`_ob`, and `_ib` for the outer and inner left, top, right, and bottom
-coordinates. The underscore is because otherwise `or` would conflict
-with a keyword.
+::: {.quirks}
+On some systems, the `measure` and `metrics` commands are awfully
```

Comment:
This is definitely true. Might be interesting to note that real browsers go to a lot of lengths to add extra caches and avoid doing work when not necessary because of this problem.